src/pydicom/pixels/decoders/base.py

```diff
@@ -0,0 +1,1911 @@
+                if (end_offset := start_offset + expected_length) > len(src):
+                    raise ValueError(
+                        f"There is insufficient pixel data to contain {index + 1} frames"
+                    )
+
+                if expected_length % 2 == 0:
+                    # Even length frame: start and end correct
+                    return src[start_offset:end_offset]
+
+                if index % 2 == 0:
+                    # Odd length frame, even index: start correct, end incorrect
+                    return src[start_offset : end_offset + 1]
+
+                # Odd length frame, odd index: start incorrect, end correct
+                return src[start_offset - 1 : end_offset]
+
+            # Return all frames
+            return src
+
+        if index is not None:
+            start_offset = index * expected_length
+            if (end_offset := start_offset + expected_length) > len(src):
+                raise ValueError(
+                    f"There is insufficient pixel data to contain {index + 1} frames"
+                )
+
+            return src[start_offset:end_offset]
+
+        expected_length *= runner.number_of_frames
+        return src[:expected_length]
+
+    @staticmethod
+    def _as_buffer_encapsulated(
+        runner: DecodeRunner, index: int | None
+    ) -> bytes | bytearray:
+        """ "Return the raw decoded pixel data as a buffer-like.
+
+        Parameters
+        ----------
+        runner : pydicom.pixels.decoders.base.DecodeRunner
+            The runner with the encoded data and decoding options.
+        index : int | None
+            The index of the frame to be returned, or ``None`` if all frames
+            are to be returned
+
+        Returns
+        -------
+        bytes | bytearray
+            A buffer-like containing the decoded pixel data.
+        """
+        length_bytes = runner.frame_length(unit="bytes")
+
+        # Return the specified frame only
+        if index is not None:
+            frame = runner.decode(index=index)
+            if (actual := len(frame)) != length_bytes:
+                raise ValueError(
+                    "Unexpected number of bytes in the decoded frame with index "
+                    f"{index} ({actual} bytes actual vs {length_bytes} expected)"
+                )
+
+            return frame
+
+        # Return all frames
+        # Preallocate container array for the frames
+        buffer = bytearray(length_bytes * runner.number_of_frames)
+        for index, frame in enumerate(runner.iter_decode()):
+            start = index * length_bytes
+            if (actual := len(frame)) != length_bytes:
+                raise ValueError(
+                    "Unexpected number of bytes in the decoded frame with index "
+                    f"{index} ({actual} bytes actual vs {length_bytes} expected)"
+                )
+
+            buffer[start : start + length_bytes] = frame
```

Comment:
Sames as for the array case, first iterate through range(number_of_frames) then concatenate any excess